vllm/worker/worker.py

```diff
@@ -263,37 +252,63 @@ def execute_model(
         blocks_to_copy = torch.tensor(execute_model_req.blocks_to_copy,
                                       device=self.device,
                                       dtype=torch.int64).view(-1, 2)
-        data: Dict[str, Any] = {
-            "num_seq_groups": num_seq_groups,
-            "blocks_to_swap_in": blocks_to_swap_in,
-            "blocks_to_swap_out": blocks_to_swap_out,
-            "blocks_to_copy": blocks_to_copy,
-        }
-        broadcast_tensor_dict(data, src=0)
 
-        self.cache_swap(blocks_to_swap_in, blocks_to_swap_out, blocks_to_copy)
+        return model_input.replace(
+            num_seq_groups=num_seq_groups,
+            blocks_to_swap_in=blocks_to_swap_in,
+            blocks_to_swap_out=blocks_to_swap_out,
+            blocks_to_copy=blocks_to_copy,
+        )
+
+    @torch.inference_mode()
+    def prepare_model_input(
+        self,
+        execute_model_req: Optional[ExecuteModelRequest] = None
+    ) -> Optional[GPUModelInput]:
+        if self.is_driver_worker:
+            if execute_model_req is None:
+                if self.parallel_config.tensor_parallel_size > 1:
+                    # This signals that there's no more requests to process for
+                    # now. All workers are running infinite loop with
+                    # broadcast_tensor_dict, and it stops the loop when the
+                    # driver broadcasts an empty input. Send an empty input to
+                    # notify all other workers to stop their execution loop.
+                    broadcast_tensor_dict({}, src=0)
+                return None
+
+            model_input = self.prepare_model_input_local(execute_model_req)
+            if self.parallel_config.tensor_parallel_size > 1:
+                metadata_dict = model_input.as_broadcastable_tensor_dict()
+                broadcast_tensor_dict(metadata_dict, src=0)
+        else:
+            metadata_dict = broadcast_tensor_dict(src=0)
+            if not metadata_dict:
+                return None
+
+            model_input = self.model_runner.get_empty_model_input()
+            model_input = model_input.new(
+                attn_backend=self.model_runner.attn_backend, **metadata_dict)
+        return model_input
+
+    @torch.inference_mode()
+    @disable_communication
+    def execute_model_local(
+        self, model_input: GPUModelInput
+    ) -> List[Union[SamplerOutput, PoolerOutput]]:
+        self.cache_swap(model_input.blocks_to_swap_in,
+                        model_input.blocks_to_swap_out,
+                        model_input.blocks_to_copy)
 
         # If there is no input, we don't need to execute the model.
-        if num_seq_groups == 0:
+        if model_input.num_seq_groups == 0:
             return []
 
-        output = self.model_runner.execute_model(seq_group_metadata_list,
-                                                 self.gpu_cache)
+        output = self.model_runner.execute_model(model_input, self.gpu_cache)
```

Comment:
this function is decorated with `@disable_communication`, then what will happen when we have TP enabled?

Reply:
This is only referring to control plane communication, so it should not affect TP execution (unless that is somehow calling broadcast_tensor_dict).

Reply:
This is kind of misleading. I don't think we need to enforce it with such a heavy mechanism (and also slight runtime overhead). Just keep it in docstring is okay. Very few people will touch these functions or call `broadcast_tensor_dict` themselves.

Reply:
How about renaming it to `forbid_control_plane_communication` or `worker_local`? And can you say more what your concerns are? The run-time overhead for the check is negligible.

Reply:
I think it is fine to just document that we don't call `broadcast_tensor_dict` in this function. We don't need to enforce it with runtime check.

Actually, after the refactor, if we don't call `broadcast_tensor_dict` in this function, and this function only deals with input that is already broadcast, who will bother adding a `broadcast_tensor_dict` here?

That's why I think this `@disable_communication` is unnecessary.